WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductFromImage/AddProductFromImageCoordinator.swift

```diff
@@ -85,3 +97,116 @@ private extension AddProductFromImageCoordinator {
+                    return continuation.resume(returning: nil)
+                }
+                Task { @MainActor in
+                    let image = await self.onWPMediaPickerCompletion(mediaItems: mediaItems, navigationController: formNavigationController)
+                    continuation.resume(returning: image)
+                }
+            })
+            self.mediaPickingCoordinator = mediaPickingCoordinator
+            mediaPickingCoordinator.showMediaPicker(source: source, from: formNavigationController)
+        }
+    }
+}
+
+// MARK: - Action handling for camera capture
+//
+private extension AddProductFromImageCoordinator {
+    @MainActor
+    func onCameraCaptureCompletion(asset: PHAsset?, error: Error?) async -> MediaPickerImage? {
+        guard let asset else {
+            return nil
+        }
+        return await withCheckedContinuation { continuation in
+            Task { @MainActor in
+                continuation.resume(returning: await requestImage(from: asset))
+            }
+        }
+    }
+}
+
+// MARK: Action handling for device media library picker
+//
+private extension AddProductFromImageCoordinator {
+    @MainActor
+    func onDeviceMediaLibraryPickerCompletion(assets: [PHAsset], navigationController: UINavigationController) async -> MediaPickerImage? {
+        await withCheckedContinuation { continuation in
+            let shouldAnimateMediaLibraryDismissal = assets.isEmpty
+            navigationController.dismiss(animated: shouldAnimateMediaLibraryDismissal) { [weak self] in
+                guard let self, let asset = assets.first else {
+                    return continuation.resume(returning: nil)
+                }
+                Task { @MainActor in
+                    continuation.resume(returning: await self.requestImage(from: asset))
+                }
+            }
+        }
+    }
+}
+
+// MARK: - Action handling for WordPress Media Library
+//
+private extension AddProductFromImageCoordinator {
+    @MainActor
+    func onWPMediaPickerCompletion(mediaItems: [Media], navigationController: UINavigationController) async -> MediaPickerImage? {
+        await withCheckedContinuation { continuation in
+            let shouldAnimateMediaLibraryDismissal = mediaItems.isEmpty
+            navigationController.dismiss(animated: shouldAnimateMediaLibraryDismissal) { [weak self] in
+                guard let self, let media = mediaItems.first else {
+                    return continuation.resume(returning: nil)
+                }
+                let productImage = media.toProductImage
+                _ = self.productImageLoader.requestImage(productImage: productImage) { image in
+                    continuation.resume(returning: .init(image: image, source: .media(media: media)))
+                }
+            }
+        }
+    }
+}
+
+private extension AddProductFromImageCoordinator {
+    func requestImage(from asset: PHAsset) async -> MediaPickerImage? {
```

Comment:
We should mark this as `@MainActor` if we decide to rewrite the other methods as suggested.

Reply:
updated in https://github.com/woocommerce/woocommerce-ios/pull/10196/commits/6c884a8a54fa1c9fba64e721ca58ef7a19a474a9